Saml logout

mjattiot opened the pull request:
### Pull Request Checklist

<!-- Please read CONTRIBUTING.rst before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

Signed-off-by: Maxime Jattiot <mjattiot@opensense.fr>

richvdh commented:
@mjattiot thanks for the contribution... first step is to merge recent `develop` into your branch so that the CI works.

mjattiot commented:
Mmmh I have just rebased from develop but I still see some checks failing. Could you help me understand why it fails ?
This is the commit on which I rebased:
commit f9f1c8acbb04ff1623ca981e76bee2d124c54858 (synapse/develop)
Merge: c01d54358 35f9165e9
Author: Erik Johnston <erik@matrix.org>
Date:   Tue Nov 26 12:56:05 2019 +0000

richvdh commented:
For example, look at the 'homeserver.log' under the artifacts tab at https://buildkite.com/matrix-dot-org/synapse/builds/5660#0ab2e7e5-c876-46dc-ab5f-bd3ad395a2f2. Looks like a syntax error of some sort.

mjattiot commented:
@richvdh the error is strange:
AttributeError: ('saml2_sp_config', "not found in ['server', 'tls', 'database', .. [https://s3.amazonaws.com/buildkiteartifacts.com/e2e088c6-a3b8-46ee-8a6a-e5fba34f6f51/fbcd2ba1-292e-4abb-9c19-deed201b5ef6/54eead4f-620f-4c83-abcb-94bade277264/9045a118-f201-49ca-93c9-08a1139ba6a5/logs/server-0/homeserver.log?response-content-type=text%2Fplain&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7L7RCW6JBA%2F20191129%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191129T205909Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEIz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJGMEQCIDYfidtBspxJTXEyCB7iaJcNxLpNYloYZsNVA%2BiweY7pAiBQgmH86F0V7hLJ5uBmKf6psV5BG9a%2FGt7T%2FgBKmEKEISrYAgjE%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDAzMjM3OTcwNTMwMyIMMkbguzeCcRuv60tgKqwCkIPnzyWPAJwzE26v02UIiRLEjfpzS9gi6aoDfclM44Ntl%2B6OyUkXJpnZ7dt%2BarT%2FTJr1LunxZGL66gan%2BSCWw9f6EqSsU6iuZ3vYleAo%2FuxSqj4Rj8LO86Pwtih%2FEQahXF5lXN%2B9q2ku4cfFOAg17d6ia9PqUcgeXLdU7U7ExXT7f91lwO724pbONEW35F2wnT%2BjcPkii3m3AtFt4Tq7EZueAHZt37pWp202JJwmkheoPpMWumQOL2WMsmmUDb50lwCPYB7%2FLhnosrG25r2KyUlKcdpBk5tNL399va%2F3tzFy%2BINjWoXQ0ohBVNI9NN5xlMN1tFBmVii9TxRyKhBJo7dlILHDz%2BJTz3ZtlzNgvY8xbXPTXi%2BcCImB%2FSAm5XwzbPRYDQwK6KJKJWP2MKPdhe8FOtECPkMVtlg51XBiJDfLiV%2FitzK2UJk2T%2BIn4%2FvrZf0sOV%2BdPn6xOHbCyqmMksM2RRmVB6jsL3tKAUfOwj2bxMop0XdAfqYmD3nH%2FmK3DB3Z6P116SfXtqVUjeXF51ZWOivgp3mkU1QVdUXExXw%2B94WLYwgkJlDQ%2FAuZfpnv%2BAY4zOU4%2F%2FhGKQNrgJ7%2F1GjICYgEfSO8nopNCunJIIZePEKuDZSra401fI46qMrWO%2FcgB8FuVMt9t3BMlIuci94K3jQj2RfwpFDeZWM3Fj4DmUHPM4tRAwYpcSJd19amdsjK9L8ejkZgzGa9LSSP7Mj5uA1Hz5xBN77IdPdx6hjgi4TE26AzC%2BWZW56oVd%2FfmhOOjfcB80pJI8YNFgdY8%2B7LNpVOCiFNV0HIvlfZFNlZi1wlYEhcSPNkWudp4aI95UXjEq1%2FTv9LCFsPJPd%2Bp7AG72GvsA%3D%3D&X-Amz-Signature=4ba84e34896396822ac06c681a25f84b66ca6d236a10632d216d57502b031b9e](url)
 
I don't understand why I am getting this error. I have not changed a single line in the saml config file. Could you help me a little bit passing this test ? Sorry I am quite new to Synapse.

richvdh commented:
The problem is the call to `get_saml_handler` in `logout.py`. This will instantiate the `SamlHandler`, which assumes that `saml2_sp_config` is set in the config file - which will not be true unless saml is enabled.

mjattiot commented:
@richvdh thank you for your insights and sorry for the response delay.
I did some changes and all tests passed ! If you have time to review some day please feel free :)

richvdh commented on synapse/rest/client/v1/logout.py:
```diff
@@ -74,6 +78,38 @@ def on_OPTIONS(self, request):
+    if hs.config.saml2_enabled:
+        SAMLLogoutServlet(hs).register(http_server)
+    else:
+        LogoutRestServlet(hs).register(http_server)
```
I'm not super-keen on having different servlet implementations bound to the same path. Can you just update the existing `LogoutRestServlet` to do the saml stuff if saml is enabled?

mjattiot replied:
It's almost the same implementation : SAMLLogoutServlet inherits LogoutRestServlet. The idea was to have no SAML logic inside LogoutRestServlet so it's cleaner. Inside SAMLLogoutServlet we re-use the whole logic with a call to self._logout() that comes from LogoutRestServlet.
At first I implemented all inside LogoutRestServlet as suggested but it doesn't fill right to mix standard logout with SAML especially if in the future we add on OpenID. Also the login servlet tried to separate that logic as well but it's easier for the login because there is a different path for sso (/redirect).
What do you think after my explanation ?

richvdh replied:
I'm afraid I still think that having a single path which can be implemented by two different servlets is inconsistent with the way everything else works in synapse. It's fine if the servlet immediately defers to one of two implementations; I just think it's hard to follow the code the way you've written it.

> Also SAMLLogoutServlet inherits LogoutRestServlet and re-uses the constructor and _logout() to share some implementations.

Yeah. I'm not a big fan of this, either. It's easy to write confusing code with inheritance.

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+        await self._auth_handler.delete_access_tokens_for_user(mxid)
+
+    # Example: https://github.com/IdentityPython/pysaml2/blob/master/example/sp-wsgi/sp.py
+    def create_logout_request(self, user, access_token):
```
please could you document the parameter types (preferably via type annotations)? Likewise for a bunch of the other new methods.

mjattiot replied:
I added extended explanation to all methods inside annotations. Please tell me if it's understandable. I am french, my english is not perfect :)

richvdh replied:
It looks like we're still missing type information for `user` and `access_token` here.

when I say "type" annotations, I mean following the style of PEP484. Writing the types down inside the docstring is ok but we're trying to switch to annotations.

```suggestion
    def create_logout_request(self, user: str, access_token, str) -> bytes:
```

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+        except Exception as e:
+            raise SynapseError(
+                500, "error while creating SAML logout request: %s" % (e,)
+            )
```
this is probably redundant?

richvdh replied:
my question is: what is the point of converting it to a SynapseError? if you let the exception propagate, the servlet wrapper will return a 500 error anyway.

On the other hand, if there are "expected" exceptions which shouldn't cause stacktraces, we should probably be returning something other than a 500.

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+        logger.info("Creating SAML logout request for %s", user)
+        try:
+            localpart = UserID.from_string(user).localpart
+            logger.debug("User localpart is %s", localpart)
```
this looks a bit superfluous

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
         # this shouldn't happen!
         raise Exception("prepare_for_authenticate didn't return a Location header")
 
+    # User should have only one local session to only one entity
```
I don't understand this sentence. what does "to only one entity" mean?

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+    # local cache. In that case we select the last session and remove
+    # other ones.
+    def _find_session_from_user(self, username):
+        subjects = self._saml_client.users.subjects()
```
how does this work when the server is restarted?

richvdh replied:
I think it's important that logout will continue to work for users who have been logged in for longer than the process has been running.

I guess we need to store the necessary information in the synapse database, against the user's device id.

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+            logger.debug("User session is %s", session)
+
+            # Creating a logout request through redirect
+            response = self._saml_client.do_logout(
```
do_logout seems to return more than one response, suggest calling this `response_map` or something

mjattiot replied:
Clarified with comments and change of the variable name

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+
+            redirect_url = next(
+                header[1] for header in http_info["headers"] if header[0] == "Location"
+            )
```
I think this is clearer:

```
        for key, value in http_info["headers"]:
            if key == "Location":
                return value
```

mjattiot replied:
Sure

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+            # The user probally logged in via m.login.password
+            if session is None:
+                return False
+            logger.debug("User session is %s", session)
```
in your `debug` lines, can you try to give more context as to what it is you are showing, so that the admin has a better chance of understanding them when they are resolving problems?

for example: `"Found SAML2 session %s"`

mjattiot replied:
Corrected

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+                "Could not logout from SAML: %s" % (resp_saml.response.status.message,),
+            )
+        except Exception as e:
+            raise SynapseError(400, "Unable to parse SAML2 response: %s" % (e,))
```
there are plenty of reasons you could end up here which are nothing to do with a failure to parse the response.

mjattiot replied:
Yes, reduced the try/catch scope, similar to whats done in _map_saml_response_to_user()

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+
+    def handle_logout_response(self, request):
+        """
+            Handle an incoming LogoutResponse to /_matrix/saml2/logout
```
```suggestion
         Handle an incoming LogoutResponse to /_matrix/saml2/logout
```

richvdh replied:
could you give some context as to when this happens? AIUI it means that the IdP has processed our logout request?

mjattiot replied:
Added comments so 3 methods so we can see which one are in charge of the 5 steps of the Single Logout Profile specification.
FYI, page 33 of the spec : https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf

richvdh commented on synapse/handlers/saml_handler.py:
```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+                resp_bytes, BINDING_HTTP_REDIRECT
+            )
+            logger.info("Received SAML logout response %s", resp_saml)
+            if resp_saml.response.status.status_code.value == STATUS_SUCCESS:
```
suggest inverting this condition

mjattiot replied:
what do you mean ?
if STATUS_SUCCESS == resp_saml.response.status.status_code.value: ?

richvdh commented on synapse/rest/saml2/logout_resource.py:
```diff
@@ -0,0 +1,40 @@
+    isLeaf = 1
+
+    def __init__(self, hs):
+        super().__init__()
```
sorry; no it's not.

mjattiot commented on synapse/python_dependencies.py:
```diff
@@ -89,7 +89,7 @@
         'eliot<1.8.0;python_version<"3.5.3"',
     ],
-    "saml2": ["pysaml2>=4.5.0"],
+    "saml2": ["pysaml2>=4.9.0"],
```
no particular reason but 4.5.0 has been released in october 2017, and according to release notes, 4.9.0 don't change any API. So I assume it would have been better to bump the minor version for stability improvement. Up to you, I can revert

richvdh replied:
generally we don't bump our minimum required dependencies unless there is a need to solve a particular problem. Bumping the version for no reason makes things difficult for distribution package maintainers.

mjattiot commented on synapse/rest/client/v1/logout.py:
```diff
@@ -74,6 +78,38 @@ def on_OPTIONS(self, request):
+            finish_request(request)
+            # Returning None because the response is handle by _saml_handler.logout
+            # instead of the synapse.http.HttpServer interface with JsonResource.
+            # See server.py for more information.
```
Corrected

richvdh commented:
this is bitrotting rapidly. I'm going to close it for now. we'd still welcome contributions in this area; if you'd like to resume work we can reopen.
